### PR TITLE
Test program for RP2040.

### PR DESCRIPTION
**Pull request by thejpster**
Provides a UART interface that lets you read and write registers, and loop a 48 kHz 16-bit stereo audio sample.

### CODE REVIEW

**Comment by thejpster** [comment]
We should wait for the nau88c22 driver release, which is waiting on Embedded HAL 1.0.

File: firmware/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,1945 @@
...
+    codec
+        .modify_clockcontrol1(|mut w| {
+            // MCLK pin is clock source, not PLL
+            w.clkm_set(false);
+            // divide MCLK by 1 to get Fs = 48 kHz x 256 = 12.288 MHz
+            w.mclksel_set(0);
+            // lower the BCLK, 2 => BCLK = Fs ÷ 4 => BCLK = 3.072 MHz
+            w.bclksel_set(2);
+            // Set to be clock output (so called 'Master' mode)
+            w.clkioen_set(true);
```
Comment (jannic) [inline firmware/src/main.rs:265]: What's the reason you let the codec provide the clock?
If you generated the clock from the PIO state machine, you'd have less latency issues, as you could write the clock signals using side-set at the exact same time as the data bits.

Comment (thejpster) [reply firmware/src/main.rs:265]: This CODEC will run with FS and  BCLK as inputs, but it has to be synchronous with the CODEC's internal IMCLK clock signal. My early Python code had the 2040 output BCLK and the audio would drift in and out (and had a ton of noise) as the RP2040 crystal drifted relative to the CODEC crystal.

Instead you have to also feed BCLK (or some other synchronous clock source) into the MCLK pin and then use the PLL to reconstruct the full Fs x 256 (i.e. 12.288 MHz for 48 kHz Fs) from MCLK, which is then in sync with the BCLK. This CODEC can do this but it involves PLL programming and I haven't written the code yet. I also don't know what happens if the BCLK stops (like if the 2040 reboots or gets reflashed) and how well the chip recovers from losing its clock. You could have a separate pin carry MCLK from the 2040 to the CODEC (pretty sure you can route clock signals to pins) but I have run out of pins.

Plus, most CODECs cannot do this PLL trick and in general, letting the CODEC generate the BCLK gives us way more options in the future. It's how I currently run the TLV320AIC23 on the Neotron Pico, having started out with it as the BCLK source.

Having said all this I guess PLL support will be required eventually so I can support 44.1 kHz Fs from my 12.288 MHz crystal.

Comment (jannic) [reply firmware/src/main.rs:265]: Ok, so it's mainly to save a GPIO pin for MCLK. Which is a good reason :-)

The only other codec I looked at was https://files.waveshare.com/upload/4/40/PCM5101A.pdf where you can just omit providing MCLK: "The device starts up expecting an external SCK input, but if BCK and LRCK start correctly while SCK remains at ground level for 16 successive LRCK periods, then the internal PLL starts, automatically generating an internal SCK from the BCK reference."
(Don't be confused by the Chinese text on the cover page of that data sheet. There are also English pages in the same PDF.)

Having a driver that can work with clocks provided by the codec (or some other source) is of course a good thing to have. In a professional environment, you might even have a separate device providing a master clock for your whole audio setup.

Comment (thejpster) [reply firmware/src/main.rs:265]: The PCM510x would work for sure, but it has a single line-output and no headphone amplifier or volume control (https://github.com/Neotron-Compute/Neotron-Pico/issues/174#issuecomment-1666548247)

**Comment by thejpster** [comment]
Oh, we also need eh1.0 in the rp2040 HAL implementation

**Pull request merged**